### PR TITLE
Use maybe_body_owned_by for multiple suggestions

### PR DESCRIPTION
**Pull request by chenyukang**
This is a continued work from https://github.com/rust-lang/rust/pull/113567

We have several other suggestions not working for closure, this PR use `maybe_body_owned_by` to fix them and add test cases for them.

### CODE REVIEW

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 1]
```diff
@@ -363,11 +363,8 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
         }
         let hir = self.infcx.tcx.hir();
-        if let Some(hir::Node::Item(hir::Item {
-            kind: hir::ItemKind::Fn(_, _, body_id),
-            ..
-        })) = hir.find(self.mir_hir_id())
-            && let Some(hir::Node::Expr(expr)) = hir.find(body_id.hir_id)
+        if let Some(body_id) = hir.maybe_body_owned_by(self.mir_def_id()) &&
+            let Some(hir::Node::Expr(expr)) = hir.find(body_id.hir_id)
         {
```
Comment (cjgillot) [inline compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:368]: A bit more canonical:
```suggestion
        if let Some(body_id) = hir.maybe_body_owned_by(self.mir_def_id()) {
            let expr = hir.body(body_id).value;
```

File: tests/ui/borrowck/copy-suggestion-region-vid.stderr [hunk 2]
```diff
@@ -8,6 +8,11 @@ LL |         HelperStruct { helpers, is_empty: helpers[0].is_empty() }
    |                        -------            ^^^^^^^^^^ value borrowed here after move
    |                        |
    |                        value moved here
+   |
+help: consider cloning the value if the performance cost is acceptable
+   |
+LL |         HelperStruct { helpers.clone(), is_empty: helpers[0].is_empty() }
```
Comment (cjgillot) [inline tests/ui/borrowck/copy-suggestion-region-vid.stderr:14]: This suggestion is wrong.

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 3]
```diff
@@ -1098,10 +1084,8 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
                 }
                 let hir_map = self.infcx.tcx.hir();
                 let def_id = self.body.source.def_id();
-                let hir_id = hir_map.local_def_id_to_hir_id(def_id.expect_local());
-                let node = hir_map.find(hir_id);
-                let hir_id = if let Some(hir::Node::Item(item)) = node
-                && let hir::ItemKind::Fn(.., body_id) = item.kind
+                let hir_id = if let Some(local_def_id) = def_id.as_local() &&
+                    let Some(body_id) = hir_map.maybe_body_owned_by(local_def_id)
             {
```
Comment (cjgillot) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:1089]: Drive-by: could you fix the indentation of this block and the one below?

File: tests/ui/borrowck/copy-suggestion-region-vid.stderr (hunk 2 above)
Comment (chenyukang) [reply tests/ui/borrowck/copy-suggestion-region-vid.stderr:14]: emm, need to take care of shorthand, let me fix it.
